src/nbl/video/CVulkanLogicalDevice.cpp

```diff
@@ -0,0 +1,149 @@
+
+    return allocateGPUMemory(additionalReqs);
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateSpilloverMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (additionalReqs.memoryHeapLocation == IDriverMemoryAllocation::ESMT_DEVICE_LOCAL)
+        return nullptr;
+
+    IDriverMemoryBacked::SDriverMemoryRequirements memoryReqs = getSpilloverGPUMemoryReqs();
+    memoryReqs.vulkanReqs.alignment = core::max(memoryReqs.vulkanReqs.alignment, additionalReqs.vulkanReqs.alignment);
+    memoryReqs.vulkanReqs.size = core::max(memoryReqs.vulkanReqs.size, additionalReqs.vulkanReqs.size);
+    memoryReqs.vulkanReqs.memoryTypeBits &= additionalReqs.vulkanReqs.memoryTypeBits;
+    memoryReqs.mappingCapability = additionalReqs.mappingCapability;
+    memoryReqs.memoryHeapLocation = additionalReqs.memoryHeapLocation;
+    memoryReqs.prefersDedicatedAllocation = additionalReqs.prefersDedicatedAllocation;
+    memoryReqs.requiresDedicatedAllocation = additionalReqs.requiresDedicatedAllocation;
+
+    return allocateGPUMemory(memoryReqs);
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateUpStreamingMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getUpStreamingMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    IDriverMemoryBacked::SDriverMemoryRequirements memoryReqs = getUpStreamingMemoryReqs();
+    memoryReqs.vulkanReqs.alignment = core::max(memoryReqs.vulkanReqs.alignment, additionalReqs.vulkanReqs.alignment);
+    memoryReqs.vulkanReqs.size = core::max(memoryReqs.vulkanReqs.size, additionalReqs.vulkanReqs.size);
+    memoryReqs.vulkanReqs.memoryTypeBits &= additionalReqs.vulkanReqs.memoryTypeBits;
+    memoryReqs.mappingCapability = additionalReqs.mappingCapability;
+    memoryReqs.memoryHeapLocation = additionalReqs.memoryHeapLocation;
+    memoryReqs.prefersDedicatedAllocation = additionalReqs.prefersDedicatedAllocation;
+    memoryReqs.requiresDedicatedAllocation = additionalReqs.requiresDedicatedAllocation;
+
+    return allocateGPUMemory(memoryReqs);
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateDownStreamingMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getDownStreamingMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    IDriverMemoryBacked::SDriverMemoryRequirements memoryReqs = getDownStreamingMemoryReqs();
+    memoryReqs.vulkanReqs.alignment = core::max(memoryReqs.vulkanReqs.alignment, additionalReqs.vulkanReqs.alignment);
+    memoryReqs.vulkanReqs.size = core::max(memoryReqs.vulkanReqs.size, additionalReqs.vulkanReqs.size);
+    memoryReqs.vulkanReqs.memoryTypeBits &= additionalReqs.vulkanReqs.memoryTypeBits;
+    memoryReqs.mappingCapability = additionalReqs.mappingCapability;
```

Comment:
READ only mapping capability

COHERENT if possible (`memoryTypeBits` exist that support it)